src/persist-txn/src/txns.rs

```diff
@@ -204,65 +202,61 @@ where
+                let data_id = data_write.shard_id();
+                (data_id, C::encode(crate::TxnsEntry::Register(data_id)))
+            })
+            .collect::<Vec<_>>();
+        let data_ids_debug = || {
+            data_writes
+                .iter()
+                .map(|x| format!("{:.9}", x.shard_id().to_string()))
+                .collect::<Vec<_>>()
+                .join(" ")
+        };
 
-        // SUBTLE! We have to write the registration to the txns shard *before*
-        // we write empty data to the physical shard. Otherwise we could race
-        // with a commit at the same timestamp, which would then not be able to
-        // copy the committed batch into the data shard.
         let mut txns_upper = self
             .txns_write
             .shared_upper()
             .into_option()
             .expect("txns should not be closed");
         loop {
             self.txns_cache.update_ge(&txns_upper).await;
-            data_ids.retain(|data_id| {
-                let Ok(ts) = self.txns_cache.data_since(data_id) else {
-                    return true;
-                };
-                debug!(
-                    "txns register {:.9} already done at {:?}",
-                    data_id.to_string(),
-                    ts
-                );
-                if max_registered_ts < ts {
-                    max_registered_ts = ts;
-                }
-                false
-            });
-            if data_ids.is_empty() {
-                register_ts = max_registered_ts;
-                break;
-            } else if register_ts < txns_upper {
-                debug!(
-                    "txns register {} at {:?} mismatch current={:?}",
-                    data_ids_debug, register_ts, txns_upper,
-                );
-                return Err(txns_upper);
-            }
-
-            let updates = data_ids
-                .iter()
-                .map(|data_id| C::encode(crate::TxnsEntry::Register(*data_id)))
-                .collect::<Vec<_>>();
+            // Figure out which are still unregistered as of `txns_upper`. Below
+            // we write conditionally on the upper being what we expect so than
+            // we can re-run this if anything changes from underneath us.
             let updates = updates
                 .iter()
-                .map(|(key, val)| ((key, val), &register_ts, 1i64))
+                .flat_map(|(data_id, (key, val))| {
+                    let registered = self.txns_cache.registered_at(data_id, &txns_upper);
+                    (!registered).then_some(((key, val), &register_ts, 1))
+                })
                 .collect::<Vec<_>>();
+            // If the txns_upper has passed register_ts, we can no longer write.
+            // Return success if we have nothing left to write or an Error if we
+            // do.
+            if register_ts < txns_upper {
+                if updates.is_empty() {
+                    break;
```

Comment:
I'm a little nervous about this. `updates` contains the shards that have not been registered at `txns_upper`, however the guarantee of this method is that all shards are registered at `register_ts`. If a shard was registered at a timestamps `ts` s.t. `register_ts < ts < txn_upper` I think we are breaking that guarantee.

I'm not sure if that matters at all? Still I wanted to point it out.

Reply:
No, I agree with you. This has been making me nervous, too. I think we just remove this Ok case entirely

Reply:
I think you could re-filter updates based on `register_ts` if you wanted to, but returning an error also seems fine.

Reply:
Ah, yeah, that's an interesting idea, but I think I'm happy with the Err. Persist-txn already has enough inherent subtlety that it's probably worth taking opportunities to be obvious when we can, and this one seems safe enough perf-wise